slack_bolt/adapter/falcon/async_resource.py

```diff
@@ -0,0 +1,84 @@
+from datetime import datetime  # type: ignore
+from http import HTTPStatus
+
+from falcon import version as falcon_version
+from falcon.asgi import Request, Response
+from slack_bolt import BoltResponse
+from slack_bolt.async_app import AsyncApp
+from slack_bolt.oauth.async_oauth_flow import AsyncOAuthFlow
+from slack_bolt.request.async_request import AsyncBoltRequest
+
+
+class AsyncSlackAppResource:
+    """
+    For use with ASGI Falcon Apps.
+
+    from slack_bolt.async_app import AsyncApp
+    app = AsyncApp()
+
+    import falcon
+    app = falcon.asgi.App()
+    app.add_route("/slack/events", AsyncSlackAppResource(app))
+    """
+
+    def __init__(self, app: AsyncApp):  # type: ignore
+        self.app = app
+
+    async def on_get(self, req: Request, resp: Response):
+        if self.app.oauth_flow is not None:
+            oauth_flow: AsyncOAuthFlow = self.app.oauth_flow
+            if req.path == oauth_flow.install_path:
+                bolt_resp = await oauth_flow.handle_installation(
+                    await self._to_bolt_request(req)
+                )
+                await self._write_response(bolt_resp, resp)
+                return
+            elif req.path == oauth_flow.redirect_uri_path:
+                bolt_resp = await oauth_flow.handle_callback(
+                    await self._to_bolt_request(req)
+                )
+                await self._write_response(bolt_resp, resp)
+                return
+
+        resp.status = "404"
+        resp.body = "The page is not found..."
+
+    async def on_post(self, req: Request, resp: Response):
+        bolt_req = await self._to_bolt_request(req)
+        bolt_resp = await self.app.async_dispatch(bolt_req)
+        await self._write_response(bolt_resp, resp)
+
+    async def _to_bolt_request(self, req: Request) -> AsyncBoltRequest:
+        return AsyncBoltRequest(
+            body=(await req.stream.read(req.content_length or 0)).decode("utf-8"),
+            query=req.query_string,
+            headers={k.lower(): v for k, v in req.headers.items()},
+        )
+
+    async def _write_response(self, bolt_resp: BoltResponse, resp: Response):
+        if falcon_version.__version__.startswith("2."):
```

Comment:
Can we move this validation in the constructor instead?

The if/else in the sync adapter is added to deal with the Falcon's API changes in major versions (`resp.body` in v2 vs `resp.text` in v3). As this ASGI version supports only v3, we can simply have `resp.text = bolt_resp.body` only in this adapter.

That being said, this version check is still good for better developer experience! I think that it can be done earlier than the timing that your app receive requests from Slack.

Reply:
Done in https://github.com/slackapi/bolt-python/pull/614/commits/515345c42aa1ca12034650f5d231829d4d27adf5